### PR TITLE
Make storage safe

### PR DESCRIPTION
**Pull request by arroyc**
Following are the changes:
1. Moved storage account config to Jenkins credential store which is more secure the original xml file  for storage config.
2. Upgrading plugin to 0.3.3 or future from an old version won't break any existing jobs.
3.  Storage Accounts can be added to credential store via credentials page from Jenkins dashboard, system configuration page as previous and job configuration page (new).

### CODE REVIEW

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 1]
```diff
@@ -81,14 +93,16 @@ public DownloadType(final String value, final String containerName, final String
...
             final DownloadType downloadType,
             final String includeFilesPattern,
             final String excludeFilesPattern,
             final String downloadDirLoc,
             final boolean flattenDirectories,
             final boolean includeArchiveZips) {
-        this.storageAccName = storageAccName;
+        this.storageCredentialId = storageCredentialId;
+        this.storageCreds = AzureCredentials.getStorageAccountCredential(storageCredentialId);
+        this.storageAccName = storageCreds.getStorageAccountName();
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: storageAccName should be transient (it's not part of the state anymore)

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 2]
```diff
@@ -153,223 +167,229 @@ public BuildStepMonitor getRequiredMonitorService() {
...
-		    run.setResult(Result.UNSTABLE);
-	    } else {
-		    listener.getLogger().println(Messages.AzureStorageBuilder_files_downloaded_count(filesDownloaded));
-	    }
-
-	} catch (Exception e) {
-	    e.printStackTrace(listener.error(Messages.AzureStorageBuilder_download_err(strAcc.getStorageAccName())));
-	    run.setResult(Result.UNSTABLE);
-	}
+        StorageAccountInfo strAcc = null;
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: in the future please don't change whitespaces in the same commit with actual changes. It's impossible to figure out what exactly was modified

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 3]
```diff
@@ -114,118 +128,138 @@
...
+        this.virtualPath = virtualPath.trim();
+        this.cleanUpContainer = cleanUpContainer;
+        this.allowAnonymousAccess = allowAnonymousAccess;
+        this.uploadArtifactsOnlyIfSuccessful = uploadArtifactsOnlyIfSuccessful;
+        this.doNotFailIfArchivingReturnsNothing = doNotFailIfArchivingReturnsNothing;
+        this.doNotUploadIndividualFiles = doNotUploadIndividualFiles;
+        this.uploadZips = uploadZips;
+        this.doNotWaitForPreviousBuild = doNotWaitForPreviousBuild;
+        this.storageCredentialId = storageCredentialId;
+        this.storageAccName = this.storageCreds.getStorageAccountName();
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: same thing as in Builder. storageAccName should be transient or completely removed (you can get it from the credential)

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder/config.jelly [hunk 4]
```diff
@@ -1,9 +1,13 @@
 <?jelly escape-by-default='true'?>
-<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:ab="/lib/credentials" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+<!--
```
Comment (clguimanMSFT) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder/config.jelly]: why leave this in a comment?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 5]
```diff
@@ -158,6 +161,35 @@ public BuildStepMonitor getRequiredMonitorService() {
         return BuildStepMonitor.NONE;
     }
 
+    public AzureCredentials.StorageAccountCredential getStorageCreds() {
+        String credentialsId = this.storageCredentialId == null
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: why would the storageCredentialId would be null but storageCreds not?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 6]
```diff
@@ -216,12 +223,32 @@ public String getStorageCredentialId() {
     }
 
     public AzureCredentials.StorageAccountCredential getStorageCreds() {
-
-        if (storageCreds == null && storageCredentialId != null) {
-            return AzureCredentials.getStorageAccountCredential(this.storageCredentialId);
+        String credentialsId = this.storageCredentialId == null
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: this method looks very similar with the one in builder

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 7]
```diff
@@ -216,12 +223,32 @@ public String getStorageCredentialId() {
...
+                    this.storageCreds = credentials;
+                    return credentials;
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
         }
-
-        return storageCreds;
+        if (this.storageCreds == null) {
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: I don't understand what you're trying to acomplish. Should this method just call into AzureCredentials.getStorageAccountCredentials(credentialsId) ?
getStorageAccountCredentials shouldn't throw an exception and just an empty credential

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java (hunk 6 above)
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: yes they are similar ... I need to find a way to clean it up

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java (hunk 7 above)
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: I think I can simplify this ..  Initially I started writing this piece of code while following the migration examples of other plugins, but now I'm migrating differently, so I can modify accordingly

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 5 above)
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: only for debugging ... needs to be removed and simplified .. will do it

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 2 above)
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: sure .. sorry for the pain .. I will make sure ... in the mean time please let me know if you find it difficult ... only major changes were in storagepublisher, storagebuilder and utils. publisher and builder are affected inside perform and databound constructors and inside dofill methods

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.properties [hunk 8]
```diff
@@ -1,3 +1,10 @@
 WindowsAzureStorage_title=Microsoft Azure Storage Account Configuration
 entry_title=Storage Account in the System (Storage Account names) 
+rep_storageAccountName=
```
Comment (clguimanMSFT) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.properties]: what with these empty props?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java [hunk 9]
```diff
@@ -118,6 +144,191 @@ public static String getBlobEP(String blobURL) {
...
+        } catch (java.security.NoSuchAlgorithmException e) {
+        }
+        return null;
+    }
+
+    public static String getWorkDirectory() {
+        return Jenkins.getInstance().root.getAbsolutePath();
     }
+
+    public static List<StorageAccountInfo> getOldStorageConfig(String fileName) throws SAXException, IOException, ParserConfigurationException {
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java]: I think the migration deserves its own class :)

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java [hunk 10]
```diff
@@ -118,6 +144,191 @@ public static String getBlobEP(String blobURL) {
...
+        }
+    }
+
+    public static void upgradeStorageConfig() throws Exception {
+        String sourceFile = getWorkDirectory() + "\\" + Constants.LEGACY_STORAGE_CONFIG_FILE;
+        try {
+            //check if it has been upgraded already
+            File upgradeFile = new File(sourceFile+".backup");
+            if(upgradeFile.exists() && upgradeFile.isFile()){
+                return;
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java]: log lines are useful around here so we can debug migration more easily

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java [hunk 11]
```diff
@@ -118,6 +144,191 @@ public static String getBlobEP(String blobURL) {
...
+        } else {
+            System.out.println("Delete operation is failed.");
+        }
+    }
+
+    public static void upgradeStorageConfig() throws Exception {
+        String sourceFile = getWorkDirectory() + "\\" + Constants.LEGACY_STORAGE_CONFIG_FILE;
+        try {
+            //check if it has been upgraded already
+            File upgradeFile = new File(sourceFile+".backup");
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java]: shouldn't the existence of a backup file mean that the previous migration attempt wasn't successful and we should try again?

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.jelly [hunk 12]
```diff
@@ -1,10 +1,10 @@
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:sf="/lib/credentials" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-
     <f:section title="${%WindowsAzureStorage_title}" id="windows-azure-storage">
         <f:entry title="${%entry_title}" >
-            <f:entry var="storageAccount" items="${descriptor.storageAccounts}" noAddButton="false" minimum="1">
+            <f:repeatable var="storageAccount" items="${descriptor.storageAccounts}" noAddButton="true" minimum="1">
```
Comment (clguimanMSFT) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.jelly]: why is this repeatable?

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.properties (hunk 8 above)
Comment (arroyc) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.properties]: removed

File: src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.jelly (hunk 12 above)
Comment (arroyc) [inline src/main/resources/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher/global.jelly]: another weird  example .. removed :)

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java (hunk 11 above)
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java]: I was using a different logic here ... updated

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java [hunk 13]
```diff
@@ -118,6 +107,30 @@ public static String getBlobEP(String blobURL) {
...
-	return DEF_BLOB_URL;
+        return Constants.DEF_BLOB_URL;
+    }
+
+    /**
+     * Returns md5 hash in string format for a given string
+     *
+     * @return
+     */
+    public static String getMD5(String md5) {
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/Utils.java]: nitpick: the input is not a md5 :)

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 14]
```diff
@@ -566,7 +613,11 @@ public ListBoxModel doFillStorageAccNameItems() {
         }
 
         public ListBoxModel doFillStorageCredentialIdItems(@AncestorInPath Item owner) {
-
+            /*if (storageAccounts.size() > 0 && azureStorageCreds.size() <= 1) {
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: probably it should be removed

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 15]
```diff
@@ -409,7 +446,17 @@ public BuildStepMonitor getRequiredMonitorService() {
     public static final class WAStorageDescriptor extends
             BuildStepDescriptor<Publisher> {
 
-        private final CopyOnWriteList<StorageAccountInfo> storageAccounts = new CopyOnWriteList<StorageAccountInfo>();
+        private static final CopyOnWriteList<StorageAccountInfo> storageAccounts = new CopyOnWriteList<StorageAccountInfo>();
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: looks like storageAccounts is an unused member?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/CredentialMigration.java [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+        FileUtils.copyFile(new File(sourceFile), new File(backupFile));
+    }
+
+    /**
+     *
+     * Take the legacy local storage credential configuration and create an
+     * equivalent global credential in Jenkins Credential Store
+     *
+     */
+    public static void removeFile(String sourceFile) throws IOException {
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/CredentialMigration.java]: nitpick: this class should have only one public method: upgradeStorageConfig

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/CredentialMigration.java [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+            // write the content into job config xml file
+            TransformerFactory transformerFactory = TransformerFactory.newInstance();
+            Transformer transformer = transformerFactory.newTransformer();
+            DOMSource source = new DOMSource(document);
+            StreamResult result = new StreamResult(new File(jobConfigFile));
+            transformer.transform(source, result);
+        }
+    }
+
+    public static void writeCredentialIdInJob(String jobName, String storageCredId) throws SAXException, IOException, ParserConfigurationException, TransformerConfigurationException, TransformerException {
```
Comment (clguimanMSFT) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/helper/CredentialMigration.java]: I wouldn't change the config files. The migration should just read the old storage accounts, but shouldn't patch jobs. The odds of something breaking are higher when you write raw bits in the config file

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java (hunk 15 above)
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: I will do another PR to refactor/removed dead codes and UNIT tests

**Pull request merged**